backend/inspirehep/records/api/literature.py

```diff
@@ -267,106 +269,149 @@ def add_files(self, data):
             LOGGER.info("Record is deleted, skipping files.", uuid=self.id)
             return data
+        builder = LiteratureBuilder()
 
         documents = data.pop("documents", [])
         figures = data.pop("figures", [])
 
-        added_documents = self.add_documents(documents)
+        self.add_files_data(documents, builder.add_document)
+        added_documents = builder.record.get("documents")
         if added_documents:
             data["documents"] = added_documents
 
-        added_figures = self.add_figures(figures)
+        self.add_files_data(figures, builder.add_figure)
+        added_figures = builder.record.get("figures")
         if added_figures:
             data["figures"] = added_figures
-
         return data
 
-    def add_documents(self, documents):
-        builder = LiteratureBuilder()
-        for document in documents:
-            if not document.get("hidden", False):
-                new_file_data = self.add_file(**document)
-                document.update(new_file_data)
-            try:
-                builder.add_document(**document)
-            except ValueError:
-                LOGGER.exception(
-                    "Duplicated document found",
-                    recid=self.get("control_number"),
-                    uuid=self.id,
-                    document=document.get("key"),
-                )
-        return builder.record.get("documents")
-
-    def add_figures(self, figures):
-        builder = LiteratureBuilder()
-        for figure in figures:
-            new_file_data = self.add_file(**figure)
-            figure.update(new_file_data)
+    def add_files_data(self, files, builder_add_file):
+        with ThreadPoolExecutor(
+            max_workers=current_app.config.get("FILES_MAX_UPLOAD_THREADS", 5)
+        ) as executor:
+            tasks = {}
+            for idx, file_data in enumerate(files):
+                if not file_data.get("hidden", False):
+                    tasks.update(
+                        {
+                            executor.submit(
+                                self.add_file,
+                                current_app.app_context(),
+                                control_number=self.get("control_number"),
+                                uuid=self.id,
+                                **file_data,
+                            ): idx
+                        }
+                    )
+                else:
+                    try:
+                        builder_add_file(**file_data)
+                    except ValueError:
+                        LOGGER.exception(
+                            "Duplicated file found",
+                            recid=self.get("control_number"),
+                            uuid=self.id,
+                            document=file_data.get("key"),
+                        )
             try:
-                builder.add_figure(**figure)
-            except ValueError:
+                for future in as_completed(
```

Comment:
do we want to handle them one by one? or just use ``wait`` instead?

Reply:
Even if we will use wait we would like to handle them one by one at the end to update values in the record. With this approach we are handling everything together easily.